lib/devices.js

```diff
@@ -78,22 +78,31 @@ module.exports = (log, db, push) => {
     return true
   }
 
-  function upsert (request, sessionToken, deviceInfo) {
+  function upsert (request, credentials, deviceInfo) {
     let operation, event, result
+    const isPlaceholderDevice = ! deviceInfo.id && ! deviceInfo.name && ! deviceInfo.type
+
     if (deviceInfo.id) {
       operation = 'updateDevice'
       event = 'device.updated'
     } else {
       operation = 'createDevice'
       event = 'device.created'
+      if (! deviceInfo.name) {
+        deviceInfo.name = 'Device'
```

Comment:
We have `synthesizeName` below [1], maybe we can expand this to take the OAuth client info as well and derive a default name from that?  (Actually just using the OAuth client name as the default name seems like it might be a fine place to start).

[1] https://github.com/mozilla/fxa-auth-server/blob/master/lib/devices.js#L135

Reply:
`synthesizeName` on create, + oauth client name

Reply:
Even if there is a case where we add the oauth client name, we either need a fallback that almost will never happen? or we should adjust the db to handle empty device names

Reply:
Is it already possible for `synthesizeName` to return an empty string, if the sessionToken was used with a user-agent string that failed to parse properly?

Reply:
Yeap

Reply:
What happens today in this case?

Reply:
The function can return empty, but i think there's validation to make sure names are required?

Reply:
Apologies for butting in to this thread, I just want to make sure it hasn't misunderstood something:

> or we should adjust the db to handle empty device names

The db is already fine with empty device names (or it's meant to be, so we should consider it a bug if you've seen a problem with this). The point of `synthesizeName` was just to be a best-effort at filling those blanks in before they make it back to the front end. We still wanted to keep them as blank in the db because that reflects the real-world state that they hadn't been explicitly named.

So to be clear, `synthesizeName` is not meant to be invoked for data on the way *in* to the db, but only when it's returning back *out* from the db.

> > Is it already possible for `synthesizeName` to return an empty string, if the sessionToken was used with a user-agent string that failed to parse properly?
>
> yeap

Although it's possible, yes, in practice it only happens when a user agent string has been tampered with to an extent that we can get nothing useful from it. If a user agent string is the empty string, for example, it's pretty hard to know what to default the device name to.

> What happens today in this case?

I think there's a translated string in the content server that acts as a fallback on the client side, iirc?

Reply:
> The point of synthesizeName was just to be a best-effort at filling those blanks in before they make it back to the front end.

ah that makes sense 👍 , adjusted the code